### PR TITLE
Add Dijkstra Search

### PR DESCRIPTION
**Pull request by macanudo527**
Here is the implementation of Dijkstra Search. Let me know if I should add more tests.

### CODE REVIEW

File: src/prezzemolo/graph.py [hunk 1]
```diff
@@ -96,3 +130,13 @@ def find_shortest_path(self, start: Vertex[ValueType], end: Vertex[ValueType], r
                 return reversed(result)
             return iter(result)
         return None
+
+    def find_dijkstra_shortest_path(self, start: Vertex[ValueType], end: Vertex[ValueType], reverse: bool = True) -> Optional[Iterator[Vertex[ValueType]]]:
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: Can you rename it to `find_weighted_shortest_path`?

File: tests/test_graph.py [hunk 2]
```diff
@@ -158,6 +158,27 @@ def test_disconnected_graph(self) -> None:
         path = g1.find_shortest_path(v1, v3)
         self.assertIsNone(path)
 
+    def test_dijkstra_search(self) -> None:
```
Comment (eprbell) [inline tests/test_graph.py]: Since this is a fairly critical core algorithm I suggest adding a few more tests:
* edge case graph with only one vertex and no edges: make sure the algorithm doesn't break
* simple triangle graph (with vertexes v1, v2, v3 and edges/weights v1->v2: 1, v2->v3: 2, v1->v3: 4): find shortest path from v1 to v3
* same as previous case but with v1->v3: 2.
* same as previous case but with v1->v3: 3.
* graph with a cycle (i.e. v2 -> v3, v3-> v4, v4 -> v2)
* also worth exploring: what happens if  the graph is made out of 2 disconnected components (i.e. two "island" graphs that are not connected by any edge? Just make sure the algorithm returns sane results (e.g. empty list or similar)
* more variants on the graph you added. meaning try to find the shortest path between other vertexes in the same graph
* any other case you may want to add

File: src/prezzemolo/graph.py [hunk 3]
```diff
@@ -87,6 +88,39 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
         # start and end are type checked inside _breadth_first_search()
         return self._breadth_first_search(start, end, vertex_2_parent=None)
 
+    def _dijkstra_search(
+        self, start: Vertex[ValueType], end: Optional[Vertex[ValueType]], vertex_2_parent: Optional[Dict[Vertex[ValueType], Optional[Vertex[ValueType]]]]
+    ) -> bool:
+        distances: Dict[Vertex[ValueType], float] = {v: float("inf") for v in self.__vertexes}
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: I would call it `distance` (singular), because it refers to the distance from a node (and it returns a single value).

File: src/prezzemolo/graph.py [hunk 4]
```diff
@@ -87,6 +88,39 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
+        distances[start] = 0.0
+        priority_queue: List[Tuple[float, Vertex[ValueType]]] = [(0.0, start)]
+        predecessors: Dict[Vertex[ValueType], Optional[Vertex[ValueType]]] = {v: None for v in self.__vertexes}
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: I would call it `predecessor` (singular). Same reasoning as above.

File: src/prezzemolo/graph.py [hunk 5]
```diff
@@ -87,6 +88,39 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
         # start and end are type checked inside _breadth_first_search()
         return self._breadth_first_search(start, end, vertex_2_parent=None)
 
+    def _dijkstra_search(
+        self, start: Vertex[ValueType], end: Optional[Vertex[ValueType]], vertex_2_parent: Optional[Dict[Vertex[ValueType], Optional[Vertex[ValueType]]]]
+    ) -> bool:
+        distances: Dict[Vertex[ValueType], float] = {v: float("inf") for v in self.__vertexes}
+        distances[start] = 0.0
+        priority_queue: List[Tuple[float, Vertex[ValueType]]] = [(0.0, start)]
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: While we are using this list as a priority queue, I suggest giving this variable a name that is descriptive of its function within the algorithm: something like `remaining` or `remaining_vertexes`.

Comment (eprbell) [reply src/prezzemolo/graph.py]: Also could you use a NamedTuple instead of a Tuple? You could call it `_ShortestDistance` or something like that.

File: src/prezzemolo/graph.py [hunk 6]
```diff
@@ -87,6 +88,39 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
+            vertex_2_parent[start] = None
+
+        while priority_queue:
+            current_distance: float
+            current_vertex: Vertex[ValueType]
+            (current_distance, current_vertex) = heappop(priority_queue)
+            if current_distance > distances[current_vertex]:
+                continue
+            for neighbor in current_vertex.neighbors:
+                new_distance = current_distance + current_vertex.get_weight(neighbor)
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: I would call this `neighbor_distance` to add descriptiveness to the name.

File: src/prezzemolo/graph.py [hunk 7]
```diff
@@ -87,6 +88,39 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
+                if vertex_2_parent is not None:
+                    vertex_2_parent[neighbor] = current_vertex
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: Shouldn't this `vertex_to_parent` logic go inside the if?

File: src/prezzemolo/graph.py (hunk 5 above)
Comment (eprbell) [inline src/prezzemolo/graph.py]: Why not use an actual [priority queue](https://docs.python.org/3/library/queue.html?highlight=priority%20queue#queue.PriorityQueue) rather than a list?

File: src/prezzemolo/graph.py [hunk 8]
```diff
@@ -87,6 +88,39 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
+                if new_distance < distances[neighbor]:
+                    distances[neighbor] = new_distance
+                    predecessors[neighbor] = current_vertex
+                    heappush(priority_queue, (new_distance, neighbor))
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: Using a list it seems to me you may be pushing duplicate elements here. E.g. if neighbor was already in the list with a higher distance you may end up with two instances of it (which is probably not fatal but it would make the queue longer). This could be resolved by using a real [priority queue](https://docs.python.org/3/library/queue.html?highlight=priority%20queue#queue.PriorityQueue) instead of a list.

**Review by eprbell — CHANGES_REQUESTED** [review]
Nice! A few suggestions here and there.

File: tests/test_graph.py (hunk 2 above)
Comment (macanudo527) [reply tests/test_graph.py]: To be clear, this is a one way algorithm and cyclic so in case 3:

> same as previous case but with v1->v3: 2.

The route should still be `[v3, v2, v1]` since the link should not be reciprocal.

Comment (macanudo527) [reply tests/test_graph.py]: never mind, read that wrong. Wasted a lot of time as a result. This is making my brain hurt. haha

Comment (eprbell) [reply tests/test_graph.py]: It's true that Graph is unidirectional. Here are the expected results for cases 2 and 3:
* case 2: there are two possible paths: v1 -> v3 and v1 -> v2 -> v3. The result should be: v1 -> v2 -> v3 (because the total weight is 3 as opposed to the other path which has weight 4).
*  case 3: there are two possible paths: v1 -> v3 and v1 -> v2 -> v3. The result should be v1 -> v3 (because the total weight is 2 as opposed to the other path which has weight 3)

Comment (eprbell) [reply tests/test_graph.py]: > never mind, read that wrong. Wasted a lot of time as a result. This is making my brain hurt. haha

Haha, that happens! I recommend actually drawing the graphs on paper so that you can reason intuitively: you can both follow the algorithm step by step on a real graph and tell what the right path is.

**Comment by macanudo527** [comment]
Ok, now I have a pretty good understanding of unbound type parameters. It was really putting me in a bind for the `dataclass`. You suggested a `NamedTuple` but maybe `dataclass` is better because I need to add `__lt__` and `__gt__` to it? I'm not even sure if I can have a `NamedTuple` with an unbound type, right?

**Comment by eprbell** [comment]
> 

It should be possible to use Generics with NamedTuple (see https://stackoverflow.com/questions/50530959/generic-namedtuple-in-python-3-6), but data class is great too (and if you need gt and lt it's the way to go). I'll review soon.

File: src/prezzemolo/graph.py [hunk 9]
```diff
@@ -14,12 +14,35 @@
...
+class _ShortestDistance(Generic[ValueType]):
+    distance: float
+    vertex: Vertex[ValueType]
+
+    # __lt__ and __gt__ are needed for PriorityQueue
+    def __lt__(self, other: object) -> bool:
+        if not other:
+            return False
+        if not isinstance(other, _ShortestDistance):
+            raise TypeError(f"Operand is not class _ShortestDistance: {repr(other)}")
```
Comment (eprbell) [inline src/prezzemolo/graph.py]:
```suggestion
            raise TypeError(f"Operand is not of class _ShortestDistance: {repr(other)}")
```

File: src/prezzemolo/graph.py [hunk 10]
```diff
@@ -14,12 +14,35 @@
...
+        return self.distance < other.distance
+
+    def __gt__(self, other: object) -> bool:
+        if not other:
+            return False
+        if not isinstance(other, _ShortestDistance):
+            raise TypeError(f"Operand is not class _ShortestDistance: {repr(other)}")
+        return self.distance > other.distance
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: A simpler option would be:
```suggestion
        return not self.__lt__(other)
```

File: src/prezzemolo/graph.py [hunk 11]
```diff
@@ -87,6 +110,42 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
         return self._breadth_first_search(start, end, vertex_2_parent=None)
 
+    def _dijkstra_search(
+        self, start: Vertex[ValueType], end: Optional[Vertex[ValueType]], vertex_2_parent: Optional[Dict[Vertex[ValueType], Optional[Vertex[ValueType]]]] = None
+    ) -> bool:
+        distance: Dict[Vertex[ValueType], float] = {v: float("inf") for v in self.__vertexes}
+        distance[start] = 0.0
+        remaining_vertexes: PriorityQueue[_ShortestDistance[ValueType]] = PriorityQueue()
+        remaining_vertexes.put(_ShortestDistance(distance=0.0, vertex=start))
+        visited: Set[Vertex[ValueType]] = set()
```
Comment (eprbell) [inline src/prezzemolo/graph.py:116]: I think we don't need `visited`: we could just initialize `remaining_vertexes` with all the vertexes and their distance before the loop, and then we iterate normally removing the nearest element (as you are doing).

File: src/prezzemolo/graph.py [hunk 12]
```diff
@@ -87,6 +110,42 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
+
+            visited.add(current_vertex)
+
+            for neighbor in current_vertex.neighbors:
+                neighbor_distance = current_distance + current_vertex.get_weight(neighbor)
+                if neighbor_distance < distance[neighbor]:
+                    if vertex_2_parent is not None:
+                        vertex_2_parent[neighbor] = current_vertex
+                    distance[neighbor] = neighbor_distance
+                    remaining_vertexes.put(_ShortestDistance(distance=neighbor_distance, vertex=neighbor))
```
Comment (eprbell) [inline src/prezzemolo/graph.py:137]: No need to put anything into `remaining_vertexes` if we initialize it as described above.

File: src/prezzemolo/graph.py [hunk 13]
```diff
@@ -87,6 +110,42 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
+        distance: Dict[Vertex[ValueType], float] = {v: float("inf") for v in self.__vertexes}
+        distance[start] = 0.0
+        remaining_vertexes: PriorityQueue[_ShortestDistance[ValueType]] = PriorityQueue()
+        remaining_vertexes.put(_ShortestDistance(distance=0.0, vertex=start))
+        visited: Set[Vertex[ValueType]] = set()
+
+        while not remaining_vertexes.empty():
+            next_vertex: _ShortestDistance[ValueType] = remaining_vertexes.get()
+            current_distance: float = next_vertex.distance
+            current_vertex: Vertex[ValueType] = next_vertex.vertex
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: In these 3 lines `next_vertex` and `current_vertex` are a little confusing because one is "current", the other is "next" but they refer to different parts of the same object. I suggest renaming `next_vertex` -> `shortest_distance`.

File: src/prezzemolo/graph.py [hunk 14]
```diff
@@ -87,6 +110,42 @@ def breadth_first_search(self, start: Vertex[ValueType], end: Vertex[ValueType])
...
+                        vertex_2_parent[neighbor] = current_vertex
+                    distance[neighbor] = neighbor_distance
+                    remaining_vertexes.put(_ShortestDistance(distance=neighbor_distance, vertex=neighbor))
+
+            if current_vertex == end:
+                return True
+
+        return False
+
+    def dijkstra_search(self, start: Vertex[ValueType], end: Vertex[ValueType]) -> bool:
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: I'm thinking we probably don't need this: it doesn't offer any advantage over `breadth_first_search()`: both these functions just check if the two vertices are connected. I'm also thinking we should rename the following symbols:
* `breadth_first_search()` -> `are_connected()`
* `find_shortest_path()` -> `breadth_first_search()`
* `find_weighted_shortest_path()` -> `dijkstra()` 
* `_dijkstra_search()` -> `_dijkstra()` (there is no need to qualify it with "search": in CS when you say Dijkstra, it's understood to be his path search algorithm)

File: src/prezzemolo/graph.py [hunk 15]
```diff
@@ -96,3 +155,13 @@ def find_shortest_path(self, start: Vertex[ValueType], end: Vertex[ValueType], r
                 return reversed(result)
             return iter(result)
         return None
+
+    def find_weighted_shortest_path(self, start: Vertex[ValueType], end: Vertex[ValueType], reverse: bool = True) -> Optional[Iterator[Vertex[ValueType]]]:
+        # start and end are type checked inside _dijkstra_search()
+        vertex_2_parent: Dict[Vertex[ValueType], Optional[Vertex[ValueType]]] = {vertex: None for vertex in self.__vertexes}
```
Comment (eprbell) [inline src/prezzemolo/graph.py]: You can just leave the dict empty here.

File: tests/test_graph.py [hunk 16]
```diff
@@ -158,6 +159,93 @@ def test_disconnected_graph(self) -> None:
...
+
+        assert path
+        self.assertEqual(list(path), [test_vertexes[9], test_vertexes[6], test_vertexes[5], test_vertexes[2], test_vertexes[1]])
+
+    @classmethod
+    def _generate_graph(cls, weights: List[int], edges: List[Tuple[int, int]]) -> Dict[int, Vertex[int]]:
+        vertexes: Dict[int, Vertex[int]] = {}
+        current_vertex: Vertex[int]
+
+        for i, edge in enumerate(edges):
```
Comment (eprbell) [inline tests/test_graph.py]: Perhaps add a check that the two parameter lists are the same length.

**Review by eprbell — COMMENTED** [review]
Thanks for adding all the tests. The code is getting nicer and nicer! There's something about classic algorithms that I find oddly gratifying: it's a bit like the mental version of eating Nutella :-)

File: tests/test_graph.py [hunk 17]
```diff
@@ -158,6 +159,93 @@ def test_disconnected_graph(self) -> None:
...
+        )
+
+        g1 = Graph[int](list(test_vertexes.values()))
+        path = g1.find_weighted_shortest_path(test_vertexes[1], test_vertexes[9])
+
+        assert path
+        self.assertEqual(list(path), [test_vertexes[9], test_vertexes[6], test_vertexes[5], test_vertexes[2], test_vertexes[1]])
+
+    @classmethod
+    def _generate_graph(cls, weights: List[int], edges: List[Tuple[int, int]]) -> Dict[int, Vertex[int]]:
```
Comment (eprbell) [inline tests/test_graph.py]: I suggest returning directly a `Graph` object, without using an intermediate `Dict`

Comment (macanudo527) [reply tests/test_graph.py]: How would you find the vertexes for the searches?

File: src/prezzemolo/graph.py (hunk 11 above)
Comment (macanudo527) [reply src/prezzemolo/graph.py:116]: Distance from what? The first vertex? How would the other distances be added if we are not putting them during the loop?

File: src/prezzemolo/graph.py (hunk 15 above)
Comment (macanudo527) [reply src/prezzemolo/graph.py]: How about just shortening it to:
```
vertex_2_parent: Dict[Vertex[ValueType], Optional[Vertex[ValueType]]] = {start: None}
```
I think only the start vertex needs initialization. This line would save two lines in `_dijkstra()`

File: src/prezzemolo/graph.py (hunk 11 above)
Comment (eprbell) [reply src/prezzemolo/graph.py:116]: Distance from the start vertex, which is zero for start itself and infinite for every other vertex (it's the same information you're using to initialize `distance`.

File: src/prezzemolo/graph.py (hunk 15 above)
Comment (eprbell) [reply src/prezzemolo/graph.py]: I think `_dijkstra()` should be self-contained and so it should initialize `vertex_2_parent`: this function is just a utility on top of `_dijkstra()`.

File: tests/test_graph.py (hunk 17 above)
Comment (eprbell) [reply tests/test_graph.py]: I was thinking something along these lines: 
```
vertexes = list(graph.vertexes)
... vertexes[3] ...
```
But the order of insertion of vertexes may be different from what you expect, so it probably won't work. Let's leave it then. 

However I think this function should return a list of Vertexes, without exposing the outer code to its internal data structure. The last line could be:
```
return sorted([v for k, v in vertexes])
```

**Review by eprbell — CHANGES_REQUESTED** [review]

File: tests/test_graph.py (hunk 17 above)
Comment (macanudo527) [reply tests/test_graph.py]: This would still mean that you need sequentially numbered vertexes starting with 0. The point of the dictionary was that you can number the vertexes however you like. 

You could, in theory, number a certain zone's vertexes 901, 902, 903 for instance to make it easier to visualize. Maybe I'm just overthinking it. 

It would also require another value check in `_generate_graph()`

Comment (eprbell) [reply tests/test_graph.py]: If non-sequential vertex names were needed, one could easily build an ad-hoc dictionary instead of a list from the `vertexes` iterator: `vertexes = [v.name:v for v in graph.vertexes]`.

I thought some more about it. How about this:
* have `_generate_graph()` return a `Graph`
* create and return the `Graph` with `return Graph(sorted([v for k, v in vertexes.values()]))`
* outside `_generate_graph()` access vertexes with `vertexes = list(graph.vertexes)` or `vertexes = [v.name:v for v in graph.vertexes]` (for the non-sequential case).

Does that work?

**Comment by macanudo527** [comment]
I think I'm completely misunderstanding what you are asking for. Is something like this how you see the dijkstra function working?
```
    def _dijkstra(
        self, start: Vertex[ValueType], end: Optional[Vertex[ValueType]], vertex_2_parent: Optional[Dict[Vertex[ValueType], Optional[Vertex[ValueType]]]] = None
    ) -> bool:
        distance: Dict[Vertex[ValueType], float] = {v: float("inf") for v in self.__vertexes}
        distance[start] = 0.0
        remaining_vertexes: PriorityQueue[_ShortestDistance[ValueType]] = PriorityQueue()
        for vertex in self.__vertexes:
            if vertex == start:
                remaining_vertexes.put(_ShortestDistance(distance=0.0, vertex=start))
            else:
                remaining_vertexes.put(_ShortestDistance(distance=float("inf"), vertex=vertex))
        # visited: Set[Vertex[ValueType]] = set()
        if vertex_2_parent is not None:
            for vertex in self.__vertexes:
                vertex_2_parent[vertex] = None

        while not remaining_vertexes.empty():
            shortest_distance: _ShortestDistance[ValueType] = remaining_vertexes.get()
            current_distance: float = shortest_distance.distance
            current_vertex: Vertex[ValueType] = shortest_distance.vertex

            # if current_vertex in visited:
            #     continue

            # visited.add(current_vertex)

            for neighbor in current_vertex.neighbors:
                neighbor_distance = current_distance + current_vertex.get_weight(neighbor)
                if neighbor_distance < distance[neighbor]:
                    if vertex_2_parent is not None:
                        vertex_2_parent[neighbor] = current_vertex
                    distance[neighbor] = neighbor_distance
                    # remaining_vertexes.put(_ShortestDistance(distance=neighbor_distance, vertex=neighbor))

            if current_vertex == end:
                return True

        return False
```

If I comment out the put to `remaining_vertexes` all dijkstra tests fail with `KeyError`. If I leave it in, the islands test fails (it tries to create a path from the second island).  Is there something off about this algorithm? I'm still not sure how it would work if we aren't putting the distances on the priority queue.

**Comment by eprbell** [comment]
> I think I'm completely misunderstanding what you are asking for. Is something like this how you see the dijkstra function working?

Ouch, my bad. I hadn't considered that the distances in the priority queue would need to be updated as well inside the loop. This means your proposed implementation is fine (including visited). Again, apologies for making you spin your wheels.

File: src/prezzemolo/graph.py (hunk 11 above)
Comment (eprbell) [reply src/prezzemolo/graph.py:116]: This was my mistake: ignore this.

File: src/prezzemolo/graph.py (hunk 12 above)
Comment (eprbell) [reply src/prezzemolo/graph.py:137]: This was my mistake: ignore this.

File: tests/test_graph.py (hunk 17 above)
Comment (eprbell) [reply tests/test_graph.py]: One more suggestion: can we have the `_generate_graph()` receive one list with a ternary tuple (start, end, weight)? That reduces risk of having mismatched lists and makes it easier to "read" the graph without having to match positions. Sorry if I'm insisting on the `Graph` creation utility but I'd like to use it to add many more tests and I'd like it to be as intuitive and readable as possible.

File: tests/test_graph.py [hunk 18]
```diff
@@ -158,6 +159,93 @@ def test_disconnected_graph(self) -> None:
         path = g1.find_shortest_path(v1, v3)
         self.assertIsNone(path)
 
+    def test_dijkstra_search(self) -> None:
+        test_vertexes: Dict[int, Vertex[int]] = TestGraph._generate_graph([5, 2, 4, 1, 1, 7, 3], [(0, 1), (0, 2), (1, 3), (1, 4), (2, 1), (2, 3), (3, 4)])
+
+        g1 = Graph[int](list(test_vertexes.values()))
+        path = g1.find_weighted_shortest_path(test_vertexes[0], test_vertexes[4])
```
Comment (eprbell) [inline tests/test_graph.py]: Can you use `reverse=False`? It makes it easier to read the output path.

File: tests/test_graph.py [hunk 19]
```diff
@@ -158,6 +159,93 @@ def test_disconnected_graph(self) -> None:
         path = g1.find_shortest_path(v1, v3)
         self.assertIsNone(path)
 
+    def test_dijkstra_search(self) -> None:
```
Comment (eprbell) [inline tests/test_graph.py]: Remove `_search` from the function name.

**Review by eprbell — CHANGES_REQUESTED** [review]
Marked my incorrect comments as resolved and added a few more nits.

**Comment by macanudo527** [comment]
This looks good now unless you can think of any other tests to build. Note the new `__lt__` and `__gt__` to Vertex so that `sorted()` works. I set the locale to the States so we can have consistent output.

**Comment by macanudo527** [comment]
re: the failed test
I'm not sure what the valid locale string would be. `en_US.utf8` works for me. Maybe it's because I'm on a Japanese machine?

File: src/prezzemolo/vertex.py [hunk 20]
```diff
@@ -56,6 +56,18 @@ def __eq__(self, other: object) -> bool:
             raise TypeError(f"Operand has non-Vertex value: {repr(other)}")
         return self.name == other.name
 
+    def __lt__(self, other: object) -> bool:
+        if not other:
+            return False
+        if not isinstance(other, Vertex):
+            raise TypeError(f"Operand has non-Vertex value: {repr(other)}")
+        setlocale(LC_ALL, "en_US.utf8")
```
Comment (eprbell) [inline src/prezzemolo/vertex.py]: There are a couple of problems with using `setlocale()` here:
* `setlocale()` is typically not called in libraries, but only in app code 
* `setlocale()` should not be called multiple times with the same value (so it shouldn't be called in this method), but only once at the beginning of the app (it can be called later too if the app wants to change locale).

What happens if we don't set the locale and we just do a normal string comparison?

**Review by eprbell — COMMENTED** [review]
Looks great! I have one more comment on `setlocale()`, otherwise it's good to go.

**Comment by eprbell** [comment]
> re: the failed test I'm not sure what the valid locale string would be. `en_US.utf8` works for me. Maybe it's because I'm on a Japanese machine?

See my online comment on `setocale()`.

**Comment by eprbell** [comment]
To summarize, if you remove the locale code from vertex.py, I think we can merge this. The string comparison can simply be:
```
return self.name < other.name
```

File: src/prezzemolo/vertex.py (hunk 20 above)
Comment (macanudo527) [reply src/prezzemolo/vertex.py]: > To summarize, if you remove the locale code from vertex.py, I think we can merge this. The string comparison can simply be:
> 
> ```
> return self.name < other.name
> ```

It's all right if capital names appear before lowercase ones [because > < sort strings according to their ASCII value](https://pythonexamples.org/python-greater-than/#4)? ie.

- Apple
- Banana
- apple
- banana

I thought that wasn't intuitive. That is why I was using `strcoll(self.name, other.name)` I wanted to lock the locale because if you don't I think it will use the system default, which would be different for different regions, potentially invalidating the tests.

Comment (eprbell) [reply src/prezzemolo/vertex.py]: Yes, in ASCII capital letters are before lowercase ones. Locale cannot be set by a library, because if all libraries set the locale, then it would be impossible to understand what locale the app would be using. This is why locale setting is left to the app. Our tests only use ASCII strings with character codes < 127, so I don't think we'll have any problem.

**Comment by macanudo527** [comment]
All right. Should be good now.

**Review by eprbell — APPROVED** [review]
Looks great!

**Comment by eprbell** [comment]
I added a few more tests and uploaded prezzemolo 0.0.4 with the latest changes.

**Comment by hupptechnologies** [comment]
[Banner](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
kamalfarahani, @bntzio, @MathmoKiwi, @iadaria, @D3VinEd, @kmz, @chandyego84, @ljt1018168, @ZoneMix, @mhasan27, @xeusteerapat, @sleepdrift3r, @FortanPireva, @liys22, @sathya-kr, @Harshit933, @manasdange, @perfectlyclear, @essldd, @Armandase, @SonnyK33, @wyuanrui, @r2uhktmo, @kumonokuma, @juwonleedev, @lifa123, @frake23, @syz99, @MikuGhoul, @tmaiorca, @Despire, @asseBadiane, @eillarra, @parkermccurley, @hungrydonkey, @BobTheSoftwareDeveloper, @fnLog0, @risk1996, @runmark, @Chibuzor5050, @stevenfeixue, @sychou, @masami87, @m-keshavarz, @abkolan, @BlainMaguire, @vannamnamthong, @lfthwjx, @Mohamilr, @vnemiatyi
